### PR TITLE
Update help with env var defaults

### PR DESCRIPTION
**Pull request by KevinAnsleyJohnson**
`Key` option help information updated to include GRID_DAEMON_KEY and
CYLINDER_PATH defaults. `service_id` option help was updated to include
GRID_SERVICE_ID environment variable. URL option help was updated to
include GRID_SERVICE_ENDPOINT default.

Signed-off-by: Kevin Johnson <kevin_johnson@cargill.com>

### CODE REVIEW

File: cli/src/main.rs [hunk 1]
```diff
@@ -1170,10 +1227,6 @@ fn run() -> Result<(), CliError> {
                             .takes_value(true)
                             .help("Workflow status of this Purchase Order version"),
                     )
-                    .arg(Arg::with_name("draft").long("draft").help(
```
Comment (dplumb94) [inline cli/src/main.rs:1173]: was this meant to be in a different commit?

**Review by dplumb94 — COMMENTED** [review]
This approach does match previous implementations, but im wondering if we should use the https://docs.rs/clap/2.32.0/clap/struct.Arg.html#method.env method instead. We also wouldnt have to manually extract the env vars if we used it.

File: cli/src/main.rs (hunk 1 above)
Comment (KevinAnsleyJohnson) [reply cli/src/main.rs:1173]: My Mistake! yes - it's now updated.

**Review by vaporos — CHANGES_REQUESTED** [review]
This is PR is a dup of #729.

See the comment in #729 about general approach we've been taking with the CLI -- preferring reference readability over completeness.

**Comment by KevinAnsleyJohnson** [comment]
> This is PR is a dup of #729.
> 
> See the comment in #729 about general approach we've been taking with the CLI -- preferring reference readability over completeness.

Looking at your comment on 729. I made updates to reduce the help length, but the story I have assigned to me asked for the help to contain the environment variables (https://github.com/hyperledger/grid/issues/330). Should we update the issue, or is there a different compromise you have in mind?

**Comment by vaporos** [comment]
> This approach does match previous implementations, but im wondering if we should use the https://docs.rs/clap/2.32.0/clap/struct.Arg.html#method.env method instead. We also wouldnt have to manually extract the env vars if we used it.

This is possibly interesting, but there are a few possible negatives. The primary negative is that environment variables and tests don't mix, so if we go this route, we may be deciding to not write tests against clap code. When we have the integration tests, it might actually be desirable to call main() (essentially) of the CLI code from within the tests to get full coverage end-to-end within the tests. (Even now, we would need to implement a way to turn off environment variable handling, but that's relatively easy.) The most obvious reason is you can't have multiple tests in the same process setting the same environment variables or you will get unpredictable behavior. It is also harder to log the source of the message in output, like splinterd does, and that can be very useful if users get confused about what the actual configuration is and how it got loaded. There might also be cases where it simply doesn't work because of how things are evaluated.

Personally, my preferred pattern for this is something similar to what we have in splinterd, where we take each source individually, (cli, env, config, default, etc.) and turn it into an intermediate "partial" config representation and then merge those into the final configuration. It is very clean but has the downside of needing to update code in several files when an option is added or changed.

**Comment by vaporos** [comment]
A middle-ground here is to keep the options simple but add an ENV section with after_help():

```
ENV:
    GRID_SERVICE_ID        Specify a default value for --service-id
    GRID_SERVICE_ENDPOINT  Specify a default value for --url
```

This probably shouldn't be added when there are SUBCOMMANDS though -- because putting it after SUBCOMMANDS makes it harder to read, so this would only be for "leaf" commands.

File: cli/src/main.rs [hunk 2]
```diff
@@ -229,6 +229,13 @@ fn run() -> Result<(), CliError> {
                                 .long("wait")
                                 .takes_value(true)
                                 .help("How long to wait for transaction to be committed")
+                        )
+                        .after_help(
+                            "ENV:\n\n\
+                                CYLINDER_PATH          Path to search for private signing keys\n\
+                                GRID_DAEMON_ENDPOINT   Specifies a default value for --url\n\
+                                GRID_DEAMON_KEY        Specifies a default value for -k, --key\n\
```
Comment (dnewh) [inline cli/src/main.rs:237]: s/DEAMON/DAEMON

This also occurs in several other places in this PR

**Review by dnewh — CHANGES_REQUESTED** [review]

File: cli/src/main.rs (hunk 2 above)
Comment (KevinAnsleyJohnson) [reply cli/src/main.rs]: Thanks! It's been updated now!

**Comment by KevinAnsleyJohnson** [comment]
> I'm seeing indentation like this:
> 
> ```
> grid-agent-list
> List all agents for a given service
> 
> USAGE:
>     grid agent list [FLAGS] [OPTIONS]
> 
> FLAGS:
>     -h, --help             Prints help information
>         --line-per-role    Displays agent information for each role on it's own line. Useful when filtering by role.
>     -q, --quiet            Do not display output
>     -V, --version          Prints version information
>     -v                     Log verbosely
> 
> OPTIONS:
>     -F, --format <format>            Output format [default: human]  [possible values: human, csv]
>         --service-id <service_id>    The ID of the service the payload should be sent to; required if running on
>                                      Splinter. Format <circuit-id>::<service-id>
> 
> ENV:
> 
> GRID_DAEMON_ENDPOINT   Specifies a default value for --url
> GRID_SERVICE_ID        Specifies a default value for --service-id
> ```
> 
> expected:
> 
> ```
> grid-agent-list
> List all agents for a given service
> 
> USAGE:
>     grid agent list [FLAGS] [OPTIONS]
> 
> FLAGS:
>     -h, --help             Prints help information
>         --line-per-role    Displays agent information for each role on it's own line. Useful when filtering by role.
>     -q, --quiet            Do not display output
>     -V, --version          Prints version information
>     -v                     Log verbosely
> 
> OPTIONS:
>     -F, --format <format>            Output format [default: human]  [possible values: human, csv]
>         --service-id <service_id>    The ID of the service the payload should be sent to; required if running on
>                                      Splinter. Format <circuit-id>::<service-id>
> 
> ENV:
>     GRID_DAEMON_ENDPOINT   Specifies a default value for --url
>     GRID_SERVICE_ID        Specifies a default value for --service-id
> ```

Totally Agree. I was following splinter, but this makes sense. Updated!

**Review by dnewh — APPROVED** [review]

**Review by vaporos — APPROVED** [review]

**Pull request merged**